File: Week-3-Algorithms/ProblemSet_3/Runoff/runoff.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define MAX 9
#define MAX_VOTERS 100

// Struct para candidatos
typedef struct
{
    string name;
    int votes;
    bool eliminated;
} candidate;

// Array de candidatos
candidate candidatos[MAX];

// Matriz de preferências dos eleitores
int preferencias[MAX_VOTERS][MAX];

// Número de candidatos e votantes
int num_candidatos;
int num_votantes;

// Protótipos
bool votar(int eleitor, int rank, string nome);
bool vencedor(void);
int encontrar_min(void);
bool empate(int min);
void eliminar(int min);
void zerar_votos(void);
void realizar_votacao(void);

int main(int argc, string argv[])
{
    if (argc < 2)
    {
        printf("Uso: runoff [candidato1] [candidato2] ...\n");
        return 1;
    }

    // Inicializa candidatos
    num_candidatos = argc - 1;
    if (num_candidatos > MAX)
    {
        printf("Máximo de candidatos é %i\n", MAX);
        return 2;
    }

    for (int i = 0; i < num_candidatos; i++)
    {
        candidatos[i].name = argv[i + 1];
        candidatos[i].votes = 0;
        candidatos[i].eliminated = false;
    }

    // Recebe número de eleitores
    num_votantes = get_int("Número de eleitores: ");
    if (num_votantes > MAX_VOTERS)
    {
        printf("Máximo de eleitores é %i\n", MAX_VOTERS);
        return 3;
    }

    // Recebe as preferências
    for (int i = 0; i < num_votantes; i++)
    {
        for (int j = 0; j < num_candidatos; j++)
        {
            string nome = get_string("Votante %i - preferência %i: ", i + 1, j + 1);
            if (!votar(i, j, nome))
            {
                printf("Voto inválido.\n");
                return 4;
            }
        }
    }

    // Executa a eleição
    while (true)
    {
        realizar_votacao();

        if (vencedor())
        {
            break;
        }

        int min = encontrar_min();
        if (empate(min))
        {
            for (int i = 0; i < num_candidatos; i++)
            {
                if (!candidatos[i].eliminated)
                {
                    printf("%s\n", candidatos[i].name);
                }
            }
            break;
        }

        eliminar(min);
    }

    return 0;
}

// Registra a preferência do eleitor
bool votar(int eleitor, int rank, string nome)
{
    for (int i = 0; i < num_candidatos; i++)
    {
        if (strcmp(nome, candidatos[i].name) == 0)
        {
            preferencias[eleitor][rank] = i;
            return true;
        }
    }
    return false;
}

// Soma votos com base nas preferências
void realizar_votacao(void)
{
    zerar_votos();

    for (int i = 0; i < num_votantes; i++)
    {
        for (int j = 0; j < num_candidatos; j++)
        {
            int index = preferencias[i][j];
            if (!candidatos[index].eliminated)
            {
                candidatos[index].votes++;
                break;
            }
        }
    }
}

// Verifica se há vencedor
bool vencedor(void)
{
    for (int i = 0; i < num_candidatos; i++)
    {
        if (!candidatos[i].eliminated && candidatos[i].votes > num_votantes / 2)
        {
            printf("Vencedor: %s\n", candidatos[i].name);
            return true;
        }
    }
    return false;
}

// Retorna o menor número de votos
int encontrar_min(void)
{
    int min = num_votantes + 1;
    for (int i = 0; i < num_candidatos; i++)
    {
        if (!candidatos[i].eliminated && candidatos[i].votes < min)
        {
            min = candidatos[i].votes;
        }
    }
    return min;
}

// Verifica empate
bool empate(int min)
{
    for (int i = 0; i < num_candidatos; i++)
    {
        if (!candidatos[i].eliminated && candidatos[i].votes != min)
        {
            return false;
        }
    }
    return true;
}

// Elimina candidatos com o menor número de votos
void eliminar(int min)
{
    for (int i = 0; i < num_candidatos; i++)
    {
        if (!candidatos[i].eliminated && candidatos[i].votes == min)
        {
            candidatos[i].eliminated = true;
            printf("Candidato eliminado: %s\n", candidatos[i].name);
        }
    }
}

// Zera votos
void zerar_votos(void)
{
    for (int i = 0; i < num_candidatos; i++)
    {
        candidatos[i].votes = 0;
    }
}

File: Week-3-Algorithms/ProblemSet_3/Runoff/runoff.js
// Max number of candidates
const MAX = 9;

//Variável para contabilizar número de votos
let totalVotos = 0;

//Array for candidates
const candidatos = [];

// Candidates have name and vote count
class Candidato {
  constructor(name, votes, eliminado) {
    this.name = name;
    this.votes = votes;
    this.eliminado = eliminado = false;
  }
}

//Função para criar candidatos
const candidatoCreator = (nome_candidatos) => {
  if (nome_candidatos.length > MAX) {
    console.log("Número de candidatos invalidado para realizar a eleição");
    return;
  } else {
    nome_candidatos.forEach((nome) => {
      //Criando candidato e adicionando no array
      candidatos.push(new Candidato(nome, 0));
      return console.log(`Candidato ${nome} criado com sucesso!!`);
    });
  }
};

//Função para votar

//Recebe a lista de preferências do eleitor
const votoPorPreferencia = (preferencias) => {
  candidatos.forEach((c) => (c.votes = 0));
  for (let i = 0; i < preferencias.length; i++) {
    for (let j = 0; j < preferencias[i].length; j++) {
      let escolhaPrincipal = preferencias[i][j];

      //find é uma função callback que espera uma função, e retorna um objeto se encontrar, senão undefined
      //If verifica se o valor é nulo ou não-nulo, então para simplificar podemos passar apenas o objeto retornado por find.
      let validarEscolhido = candidatos.find(
        (c) => c.name === escolhaPrincipal && !c.eliminado
      );

      if (validarEscolhido) {
        console.log(`Voto válidado para o candidato ${validarEscolhido.name}`);
        validarEscolhido.votes++;
        totalVotos++;
        break;
      }
    }
  }
};
//Função para eliminar
const eliminarCandidato = () => {
  let menorVotoEncontrado = Infinity;

  // Passo 1: Encontrar o menor número de votos entre os não eliminados
  candidatos.forEach((c) => {
    if (!c.eliminado && c.votes < menorVotoEncontrado) {
      menorVotoEncontrado = c.votes;
    }
  });

  // Passo 2: Eliminar todos que tiverem esse menor número de votos
  candidatos.forEach((c) => {
    if (!c.eliminado && c.votes === menorVotoEncontrado) {
      c.eliminado = true;
      console.log(`Candidato ${c.name} eliminado com ${c.votes} votos.`);
    }
  });
};

//Função que verifica o resultado

const verificarResultado = (participantes) => {
  for (let c of candidatos) {
    if (!c.eliminado && c.votes > participantes.length / 2) {
      console.log(`Vencedor da eleição ${c.name}`);
      return true;
    }
  }
  return false;
};

//Função para executar  o algoritmo de Runoff/A eleição
const executarEleicao = (preferencias) => {
  while (true) {
    if (verificarResultado(preferencias)) {
      break;
    }

    eliminarCandidato();
  }
};

//Inputs para testar as funções

candidatoCreator(["Caio", "Ribamar", "Pedro"]);
const votos = [
  ["Ribamar", "Caio", "Pedro"],
  ["Ribamar", "Caio", "Pedro"],
  ["Caio", "Pedro", "Caio"],
  ["Caio", "Pedro", "Ribamar"],
  ["Pedro", "Caio", "Ribamar"],
];
executarEleicao(votos);
